### ISSUE TITLE
Service doesn't enter the "failed" state

### ISSUE THREAD

**Issue by evverx** [systemd/systemd#3166]
How to reproduce:

```
# cat <<'EOF' >/etc/systemd/system/fail-on-restart.service
[Service]
ExecStart=/bin/false
Restart=always
EOF

# systemctl daemon-reload

# systemctl start fail-on-restart

# systemctl status --no-pager -n0 fail-on-restart
● fail-on-restart.service
   Loaded: loaded (/etc/systemd/system/fail-on-restart.service; static; vendor preset: enabled)
   Active: inactive (dead)

# journalctl -u fail-on-restart --no-pager --no-hostname
-- Logs begin at Sun 2016-05-01 17:01:22 UTC, end at Sun 2016-05-01 17:03:36 UTC. --
May 01 17:03:05 systemd[1]: Started fail-on-restart.service.
May 01 17:03:05 systemd[522]: fail-on-restart.service: Executing: /bin/false
May 01 17:03:06 systemd[1]: fail-on-restart.service: Main process exited, code=exited, status=1/FAILURE
May 01 17:03:06 systemd[1]: fail-on-restart.service: Unit entered failed state.
May 01 17:03:06 systemd[1]: fail-on-restart.service: Failed with result 'exit-code'.
May 01 17:03:06 systemd[1]: fail-on-restart.service: Service hold-off time over, scheduling restart.
May 01 17:03:08 systemd[1]: Stopped fail-on-restart.service.
May 01 17:03:08 systemd[1]: Started fail-on-restart.service.
May 01 17:03:08 systemd[537]: fail-on-restart.service: Executing: /bin/false
May 01 17:03:08 systemd[1]: fail-on-restart.service: Main process exited, code=exited, status=1/FAILURE
May 01 17:03:08 systemd[1]: fail-on-restart.service: Unit entered failed state.
May 01 17:03:08 systemd[1]: fail-on-restart.service: Failed with result 'exit-code'.
May 01 17:03:08 systemd[1]: fail-on-restart.service: Service hold-off time over, scheduling restart.
May 01 17:03:08 systemd[1]: Stopped fail-on-restart.service.
May 01 17:03:08 systemd[1]: Started fail-on-restart.service.
May 01 17:03:08 systemd[542]: fail-on-restart.service: Executing: /bin/false
May 01 17:03:09 systemd[1]: fail-on-restart.service: Child 542 belongs to fail-on-restart.service
May 01 17:03:09 systemd[1]: fail-on-restart.service: Main process exited, code=exited, status=1/FAILURE
May 01 17:03:09 systemd[1]: fail-on-restart.service: Changed running -> failed
May 01 17:03:09 systemd[1]: fail-on-restart.service: Unit entered failed state.
May 01 17:03:09 systemd[1]: fail-on-restart.service: Failed with result 'exit-code'.
May 01 17:03:09 systemd[1]: fail-on-restart.service: Changed failed -> auto-restart
May 01 17:03:09 systemd[1]: fail-on-restart.service: cgroup is empty
May 01 17:03:09 systemd[1]: fail-on-restart.service: Service hold-off time over, scheduling restart.
May 01 17:03:09 systemd[1]: fail-on-restart.service: Trying to enqueue job fail-on-restart.service/restart/fail
May 01 17:03:09 systemd[1]: fail-on-restart.service: Installed new job fail-on-restart.service/restart as 339
May 01 17:03:09 systemd[1]: fail-on-restart.service: Enqueued job fail-on-restart.service/restart as 339
May 01 17:03:09 systemd[1]: fail-on-restart.service: Scheduled restart job.
May 01 17:03:09 systemd[1]: fail-on-restart.service: Changed auto-restart -> dead
May 01 17:03:09 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/restart finished, result=done
May 01 17:03:09 systemd[1]: Stopped fail-on-restart.service.
May 01 17:03:09 systemd[1]: fail-on-restart.service: Converting job fail-on-restart.service/restart -> fail-on-restart.service/start
May 01 17:03:09 systemd[1]: fail-on-restart.service: About to execute: /bin/false
May 01 17:03:09 systemd[1]: fail-on-restart.service: Forked /bin/false as 548
May 01 17:03:09 systemd[1]: fail-on-restart.service: Changed dead -> running
May 01 17:03:09 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/start finished, result=done
May 01 17:03:09 systemd[1]: Started fail-on-restart.service.
May 01 17:03:10 systemd[1]: fail-on-restart.service: Child 548 belongs to fail-on-restart.service
May 01 17:03:10 systemd[1]: fail-on-restart.service: Main process exited, code=exited, status=1/FAILURE
May 01 17:03:10 systemd[1]: fail-on-restart.service: Changed running -> failed
May 01 17:03:10 systemd[1]: fail-on-restart.service: Unit entered failed state.
May 01 17:03:10 systemd[1]: fail-on-restart.service: Failed with result 'exit-code'.
May 01 17:03:10 systemd[1]: fail-on-restart.service: Changed failed -> auto-restart
May 01 17:03:10 systemd[1]: fail-on-restart.service: cgroup is empty
May 01 17:03:10 systemd[1]: fail-on-restart.service: Service hold-off time over, scheduling restart.
May 01 17:03:10 systemd[1]: fail-on-restart.service: Trying to enqueue job fail-on-restart.service/restart/fail
May 01 17:03:10 systemd[1]: fail-on-restart.service: Installed new job fail-on-restart.service/restart as 388
May 01 17:03:10 systemd[1]: fail-on-restart.service: Enqueued job fail-on-restart.service/restart as 388
May 01 17:03:10 systemd[1]: fail-on-restart.service: Scheduled restart job.
May 01 17:03:11 systemd[1]: fail-on-restart.service: Changed auto-restart -> dead
May 01 17:03:11 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/restart finished, result=done
May 01 17:03:11 systemd[1]: Stopped fail-on-restart.service.
May 01 17:03:11 systemd[1]: fail-on-restart.service: Converting job fail-on-restart.service/restart -> fail-on-restart.service/start
May 01 17:03:11 systemd[1]: fail-on-restart.service: About to execute: /bin/false
May 01 17:03:11 systemd[1]: fail-on-restart.service: Forked /bin/false as 554
May 01 17:03:11 systemd[1]: fail-on-restart.service: Changed dead -> running
May 01 17:03:11 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/start finished, result=done
May 01 17:03:11 systemd[1]: Started fail-on-restart.service.
May 01 17:03:11 systemd[554]: fail-on-restart.service: Executing: /bin/false
May 01 17:03:11 systemd[1]: fail-on-restart.service: Child 554 belongs to fail-on-restart.service
May 01 17:03:11 systemd[1]: fail-on-restart.service: Main process exited, code=exited, status=1/FAILURE
May 01 17:03:11 systemd[1]: fail-on-restart.service: Changed running -> failed
May 01 17:03:11 systemd[1]: fail-on-restart.service: Unit entered failed state.
May 01 17:03:11 systemd[1]: fail-on-restart.service: Failed with result 'exit-code'.
May 01 17:03:11 systemd[1]: fail-on-restart.service: Changed failed -> auto-restart
May 01 17:03:11 systemd[1]: fail-on-restart.service: cgroup is empty
May 01 17:03:12 systemd[1]: fail-on-restart.service: Service hold-off time over, scheduling restart.
May 01 17:03:12 systemd[1]: fail-on-restart.service: Trying to enqueue job fail-on-restart.service/restart/fail
May 01 17:03:12 systemd[1]: fail-on-restart.service: Installed new job fail-on-restart.service/restart as 437
May 01 17:03:12 systemd[1]: fail-on-restart.service: Enqueued job fail-on-restart.service/restart as 437
May 01 17:03:12 systemd[1]: fail-on-restart.service: Scheduled restart job.
May 01 17:03:12 systemd[1]: fail-on-restart.service: Changed auto-restart -> dead
May 01 17:03:12 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/restart finished, result=done
May 01 17:03:12 systemd[1]: Stopped fail-on-restart.service.
May 01 17:03:12 systemd[1]: fail-on-restart.service: Converting job fail-on-restart.service/restart -> fail-on-restart.service/start
May 01 17:03:12 systemd[1]: fail-on-restart.service: Start request repeated too quickly.
May 01 17:03:12 systemd[1]: fail-on-restart.service: Job fail-on-restart.service/start finished, result=failed
May 01 17:03:12 systemd[1]: Failed to start fail-on-restart.service.
May 01 17:03:12 systemd[1]: fail-on-restart.service: Collecting.
May 01 17:03:27 systemd[1]: fail-on-restart.service: Collecting.
May 01 17:03:27 systemd[1]: fail-on-restart.service: Collecting.
```

This regression was introduced by 6bf0f408e4833152

```
6bf0f408e4833152197fb38fb10a9989c89f3a59 is the first bad commit
commit 6bf0f408e4833152197fb38fb10a9989c89f3a59
Author: Lennart Poettering <lennart@poettering.net>
Date:   Tue Feb 9 18:38:03 2016 +0100

    core: make the StartLimitXYZ= settings generic and apply to any kind of unit, not just services

    This moves the StartLimitBurst=, StartLimitInterval=, StartLimitAction=, RebootArgument= from the [Service] section
    into the [Unit] section of unit files, and thus support it in all unit types, not just in services.

    This way we can enforce the start limit much earlier, in particular before testing the unit conditions, so that
    repeated start-up failure due to failed conditions is also considered for the start limit logic.

    For compatibility the four options may also be configured in the [Service] section still, but we only document them in
    their new section [Unit].

    This also renamed the socket unit failure code "service-failed-permanent" into "service-start-limit-hit" to express
    more clearly what it is about, after all it's only triggered through the start limit being hit.

    Finally, the code in busname_trigger_notify() and socket_trigger_notify() is altered to become more alike.

    Fixes: #2467

:040000 040000 38f0a8f6802b535a8e292d994fdc51f477cb2d6b 2bc98593ba38556d08cce4914503216c44f2d310 M  man
:040000 040000 746fd9a37824d22d91519d1b06da867daefd8f46 15a6561c3045c885740dc4c0b744a6c497dbe2fb M  src
```

~~The test is coming soon.~~ See https://github.com/systemd/systemd/pull/3180

**Comment by evverx** [comment]
Hm...
Seems like the fix for https://github.com/systemd/systemd/issues/2467 is broken too.

6bf0f408e4833152197fb38fb10a9989c89f3a59 - `echo a | nc -U /run/test.ctl` works fine
master - `echo a | nc -U /run/test.ctl` hangs

Created: https://github.com/systemd/systemd/issues/3167

**Comment by poettering** [comment]
Fix waiting in #3170.
